### PR TITLE
grad function

### PR DESCRIPTION
**Pull request by beam2d**
This PR implements `chainer.grad` function proposed at #2871. This function implements backpropagation. The most important difference from `Variable.backward()` is that users specify the set of input variables to which the errors should be backpropagated. The resulting gradients are not accumulated to `Variable.grad_var` attribute by default, so it's easier to use when users want to compute gradients w.r.t. variables not spanning the whole parameter set. This function also skips unnecessary paths to compute gradients w.r.t. specified inputs in combination with the economical backprop API of #2970.

The usage is simple:
```py
grad_inputs = chainer.grad(outputs, inputs)
```
You can also specify the initial gradients for outputs and inputs. The initial gradients for inputs are used for gradient accumulation. If `set_grad=True` is specified, it also sets the gradients to `Variable.grad_var` of inputs like `Variable.backward()` does. The `retain_grad` option also exists.

This PR depends on #2970.

### CODE REVIEW

**Comment by unnonouno** [comment]
Can you rebase it?

**Comment by beam2d** [comment]
rebased

File: tests/chainer_tests/test_function_node.py [hunk 1]
```diff
@@ -471,4 +471,147 @@ def test_retain_outputs_f1(self):
...
+        raise NotImplementedError
+
+    def expected_grad(self):
+        raise NotImplementedError
+
+    def expected_double_grad(self):
+        raise NotImplementedError
+
+    def _print_variables(self, name, vs):
+        print('{}: '.format(name), end='')
```
Comment (mitmul) [inline tests/chainer_tests/test_function_node.py]: Please add `from __future__ import print_function` to use `end` argument for `print` in Python 2

File: chainer/function_node.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import collections
+import heapq
 import traceback
 import weakref
 
 import six
+import numpy
```
Comment (mitmul) [inline chainer/function_node.py]: L:6-7 causes a flake8 error. Please switch these lines

File: chainer/function_node.py [hunk 3]
```diff
@@ -589,3 +591,225 @@ def delete_hook(self, name):
...
+
+    This function implements the backpropagation algorithm. While
+    :meth:`Variable.backward` also implements backprop, this function selects
+    the smallest paths in the computational graph needed to compute the
+    gradients w.r.t. inputs. The error is backpropagated only through these
+    selected paths, which may reduce the overall computational cost.
+
+    This function also differs from :meth:`Variable.backward` in the way to
+    return the gradients; it directly returns the gradient variables as a list
+    instead of setting gradients to the :attr:`Variable.grad_var` attirbute of
```
Comment (mitmul) [inline chainer/function_node.py]: typo: `attirbute` -> `attribute`

File: chainer/function_node.py [hunk 4]
```diff
@@ -589,3 +591,225 @@ def delete_hook(self, name):
...
+    the original variable. It means users do not need to clear the gradient
+    w.r.t. each variable before computing the gradient using this function.
+    If ``set_grad`` option is set to ``True``, the computed gradient is also
+    stored to the :attr:`Variable.grad_var` attribute of each variable, whereas
```
Comment (mitmul) [inline chainer/function_node.py]: "stored to" -> "stored in"?

File: chainer/function_node.py [hunk 5]
```diff
@@ -589,3 +591,225 @@ def delete_hook(self, name):
...
+        grad_inputs: A sequence of variables that gives the initial value of
+            each input gradient. The gradients computed by the backprop
+            algorithm are accumulated to them (not in-place). If an element
+            is set to ``None``, the gradient is not accumulated to this value.
+            If this argument itself is ``None``, it is treated as a sequence of
+            ``None`` s.
+        set_grad (bool): If it is ``True``, the :attr:`Variable.grad_var`
+            attribute of each input is set to the computed gradient arrays.
+        retain_grad (bool): If it is ``True``, the gradients w.r.t. all the
+            intermediate variables are stored to the :attr:`Variable.grad_var`
```
Comment (mitmul) [inline chainer/function_node.py]: "stored to" -> "stored in"?

File: tests/chainer_tests/test_function_node.py (hunk 1 above)
Comment (beam2d) [inline tests/chainer_tests/test_function_node.py]: added

**Comment by beam2d** [comment]
Fixed for @mitmul's comments.

File: chainer/function_node.py [hunk 6]
```diff
@@ -589,3 +591,225 @@ def delete_hook(self, name):
 
         """
         del self.local_function_hooks[name]
+
+
+def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
```
Comment (mitmul) [inline chainer/function_node.py]: Please add this to the document. According to #3117, I think add this function to `docs/source/reference/core/function.rst` is appropriate?

Comment (beam2d) [reply chainer/function_node.py]: Oh, I forgot it. Will do. As you wrote, `function.rst` looks a good place.

**Comment by beam2d** [comment]
Fixed, but I found that `check_double_grad` tests do not run in current code. I will fix it.

**Comment by beam2d** [comment]
Fixed the tests.

File: chainer/function_node.py [hunk 7]
```diff
@@ -589,3 +591,232 @@ def delete_hook(self, name):
...
+    gradients w.r.t. inputs. The error is backpropagated only through these
+    selected paths, which may reduce the overall computational cost.
+
+    This function also differs from :meth:`Variable.backward` in the way to
+    return the gradients; it directly returns the gradient variables as a list
+    instead of setting gradients to the :attr:`Variable.grad_var` attribute of
+    the original variable. It means users do not need to clear the gradient
+    w.r.t. each variable before computing the gradient using this function.
+    If ``set_grad`` option is set to ``True``, the computed gradient is also
+    stored in the :attr:`Variable.grad_var` attribute of each variable, whereas
```
Comment (bkvogel) [inline chainer/function_node.py]: Consider rewording:

If ``set_grad`` option is set to ``True``, the computed gradient is also
    stored in the :attr:`Variable.grad_var` attribute of each variable, in which case
    any original value of :attr:`Variable.grad_var` will be updated even if it had already
    been set.

File: chainer/function_node.py [hunk 8]
```diff
@@ -589,3 +591,232 @@ def delete_hook(self, name):
...
+            each output gradient. If an element is set to ``None``, an array
+            filled with 1 is used. If this argument itself is ``None``, it is
+            treated as a sequence of ``None`` s.
+        grad_inputs: A sequence of variables that gives the initial value of
+            each input gradient. The gradients computed by the backprop
+            algorithm are accumulated to them (not in-place). If an element
+            is set to ``None``, the gradient is not accumulated to this value.
+            If this argument itself is ``None``, it is treated as a sequence of
+            ``None`` s.
+        set_grad (bool): If it is ``True``, the :attr:`Variable.grad_var`
```
Comment (bkvogel) [inline chainer/function_node.py]: Consider rewording slightly:

If it is ``True``, the :attr:`Variable.grad_var`
            attribute of each input variable is set to the corresponding computed gradient variable.

File: chainer/function_node.py [hunk 9]
```diff
@@ -589,3 +591,232 @@ def delete_hook(self, name):
...
+            attribute of each input is set to the computed gradient arrays.
+        retain_grad (bool): If it is ``True``, the gradients w.r.t. all the
+            intermediate variables are stored in the :attr:`Variable.grad_var`
+            attribute. In this case, the ``set_grad`` option is ignored.
+        enable_double_backprop (bool): If it is ``True``, the computed
+            gradients can be further backpropagated. Enabling it would increase
```
Comment (bkvogel) [inline chainer/function_node.py]: In "Enabling it would increase" consider replacing 'would' with 'will' or 'may'.

File: chainer/function_node.py (hunk 8 above)
Comment (beam2d) [reply chainer/function_node.py]: fixed.

**Comment by beam2d** [comment]
Fixed the wording in the docstring.

**Comment by mitmul** [comment]
Jenkins, test this please

File: chainer/function_node.py [hunk 10]
```diff
@@ -593,3 +595,233 @@ def delete_hook(self, name):
...
+
+
+def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
+         retain_grad=False, enable_double_backprop=False):
+    """Computes the gradient of output variables w.r.t. the input variables.
+
+    This function implements the backpropagation algorithm. While
+    :meth:`Variable.backward` also implements backprop, this function selects
+    the smallest paths in the computational graph needed to compute the
+    gradients w.r.t. inputs. The error is backpropagated only through these
```
Comment (mitmul) [inline chainer/function_node.py]: Could you replace `w.r.t. ` with `w.r.t.\\ ` (please note that one blank space is needed after double backslash `\\`)?

File: chainer/function_node.py [hunk 11]
```diff
@@ -593,3 +595,233 @@ def delete_hook(self, name):
...
+    selected paths, which may reduce the overall computational cost.
+
+    This function also differs from :meth:`Variable.backward` in the way to
+    return the gradients; it directly returns the gradient variables as a list
+    instead of setting gradients to the :attr:`Variable.grad_var` attribute of
+    the original variable. It means users do not need to clear the gradient
+    w.r.t. each variable before computing the gradient using this function.
```
Comment (mitmul) [inline chainer/function_node.py]: Could you replace `w.r.t. ` with `w.r.t.\\ ` (please note that one blank space is needed after double backslash `\\`)?

File: chainer/function_node.py [hunk 12]
```diff
@@ -593,3 +595,233 @@ def delete_hook(self, name):
             del self.local_function_hooks[name]
         else:
             raise KeyError('Hook %s does not exist' % name)
+
+
+def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
+         retain_grad=False, enable_double_backprop=False):
+    """Computes the gradient of output variables w.r.t. the input variables.
```
Comment (mitmul) [inline chainer/function_node.py]: Could you replace `w.r.t. ` with `w.r.t.\\ ` (please note that one blank space is needed after double backslash `\\`)?

File: chainer/function_node.py [hunk 13]
```diff
@@ -593,3 +595,233 @@ def delete_hook(self, name):
...
+        grad_inputs: A sequence of variables that gives the initial value of
+            each input gradient. The gradients computed by the backprop
+            algorithm are accumulated to them (not in-place). If an element
+            is set to ``None``, the gradient is not accumulated to this value.
+            If this argument itself is ``None``, it is treated as a sequence of
+            ``None`` s.
+        set_grad (bool): If it is ``True``, the :attr:`Variable.grad_var`
+            attribute of each input variable is set to the corresponding
+            computed gradient variable.
+        retain_grad (bool): If it is ``True``, the gradients w.r.t. all the
```
Comment (mitmul) [inline chainer/function_node.py]: Could you replace `w.r.t. ` with `w.r.t.\\ ` (please note that one blank space is needed after double backslash `\\`)?

File: chainer/function_node.py [hunk 14]
```diff
@@ -593,3 +595,233 @@ def delete_hook(self, name):
...
+            intermediate variables are stored in the :attr:`Variable.grad_var`
+            attribute. In this case, the ``set_grad`` option is ignored.
+        enable_double_backprop (bool): If it is ``True``, the computed
+            gradients can be further backpropagated. Enabling it may increase
+            the memory consumption (and possibly the computational time) to
+            remember the intermediate gradient values for the second
+            backpropagation.
+
+    Returns:
+        A list of gradient variables w.r.t. the inputs.
```
Comment (mitmul) [inline chainer/function_node.py]: Could you replace `w.r.t. ` with `w.r.t.\\ ` (please note that one blank space is needed after double backslash `\\`)?

**Comment by beam2d** [comment]
jenkins, test this please

**Comment by beam2d** [comment]
Jenkins, test this please.
